### PR TITLE
CallbackThrottleByTime test

### PR DESCRIPTION
**Pull request by Mecharyry**
## Problem
Wanted to test the `CallbackThrottleByTime` class but it creates the objects responsible for threading internally. 

## Solution
Create a `Scheduler` interface to hide the threading mechanism that is used from the `CallbackThrottleByTime` logic. 

Create a `SchedulerFactory` for injecting the dependencies we need to create a `FixedRateTimerScheduler` in production code and still be able to vary it for testing purposes. 

Fully test both the `CallbackThrottleByTime` and the `FixedRateTimerScheduler` class 😄

### CODE REVIEW

File: library/src/main/java/com/novoda/downloadmanager/CallbackThrottleByTime.java [hunk 1]
```diff
@@ -1,60 +1,47 @@
...
+    private final Scheduler scheduler;
 
     private DownloadBatchStatus downloadBatchStatus;
-    private Timer timer;
-    private TimerTask timerTask;
     private DownloadBatchCallback callback;
 
-    CallbackThrottleByTime(long period) {
-        this.period = period;
+    CallbackThrottleByTime(Scheduler scheduler) {
```
Comment (Hutch4) [inline library/src/main/java/com/novoda/downloadmanager/CallbackThrottleByTime.java]: 👍

File: library/src/main/java/com/novoda/downloadmanager/Scheduler.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package com.novoda.downloadmanager;
+
+interface Scheduler {
```
Comment (Hutch4) [inline library/src/main/java/com/novoda/downloadmanager/Scheduler.java]: Is Rx used in downloadmanager at all?

I'd avoid the use of Scheduler as an interface name, as it's very common. 
`DownloadActionScheduler` or `ActionScheduler` might be better?

File: library/src/main/java/com/novoda/downloadmanager/CallbackThrottleCreator.java [hunk 3]
```diff
@@ -41,7 +41,7 @@ private CallbackThrottleCreator(Type type, TimeUnit timeUnit, long frequency, Cl
     CallbackThrottle create() {
         switch (type) {
             case THROTTLE_BY_TIME:
-                return new CallbackThrottleByTime(timeUnit.toMillis(frequency));
+                return new CallbackThrottleByTime(SchedulerFactory.createFixedRateTimerScheduler(timeUnit.toMillis(frequency)));
```
Comment (Hutch4) [inline library/src/main/java/com/novoda/downloadmanager/CallbackThrottleCreator.java]: i'd make extract out the Scheduler creation into a local var. To make it clear about what is being created.

**Comment by Mecharyry** [comment]
@Hutch4 I've addressed your comments.

**Pull request merged**